rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -836,148 +823,282 @@ rmw_ret_t configure_qos_for_security(
+  //   updates and triggering a guard condition, and so that should be safe.
+  // - There might be a significant cost to calling the graph cache function to update its internal
+  //   state, and having that occur on the thread receiving data from the network may not be wise
+  // So it may be better to create a thread and a waitset.
+  if ((ret = discovery_thread_start(context)) != RMW_RET_OK) {
+    return ret;
+  }
+
+  graph_guard_condition.release();
+  publisher.release();
+  subscription.release();
+  impl.release();
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_ret_t rmw_shutdown(rmw_context_t * context)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    context,
+    context->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  // Nothing to do here for now.
+  // This is just the middleware's notification that shutdown was called.
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_ret_t rmw_context_fini(rmw_context_t * context)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    context,
+    context->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  delete context->impl;
+  *context = rmw_get_zero_initialized_context();
+  return RMW_RET_OK;
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+///////////                                                                   ///////////
+///////////    NODES                                                          ///////////
+///////////                                                                   ///////////
+/////////////////////////////////////////////////////////////////////////////////////////
+
+extern "C" rmw_ret_t rmw_node_assert_liveliness(const rmw_node_t * node)
+{
+  RET_WRONG_IMPLID(node);
+  if (dds_assert_liveliness(node->context->impl->ppant) != DDS_RETCODE_OK) {
+    return RMW_RET_ERROR;
+  }
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_node_t * rmw_create_node(
+  rmw_context_t * context, const char * name,
+  const char * namespace_, size_t domain_id,
+  bool localhost_only)
+{
+  static_cast<void>(domain_id);
+  static_cast<void>(localhost_only);
+  RET_NULL_X(name, return nullptr);
+  RET_NULL_X(namespace_, return nullptr);
+  rmw_ret_t ret;
+  int dummy_validation_result;
+  size_t dummy_invalid_index;
+  if ((ret =
+    rmw_validate_node_name(name, &dummy_validation_result, &dummy_invalid_index)) != RMW_RET_OK)
+  {
+    return nullptr;
   }
-#endif
 
   /* Since ROS2 doesn't require anything fancy from DDS Subscribers or Publishers, create a single
      pair & reuse that */
+  const dds_entity_t pp = context->impl->ppant;
   dds_entity_t pub, sub;
   if ((pub = dds_create_publisher(pp, nullptr, nullptr)) < 0) {
-    node_gone_from_domain_locked(did);
```

Comment:
just out of curiosity, what are these `publisher` and `subscription` for?

Reply:
These are the DDS publisher and subscriber objects underneath which the individual writers and readers are created. As there is nothing related to the node in them (the "group data" setting would be about the only thing that could've been used to some purpose) there is no benefit to having a pair of these per node. I've moved them to the context.

Reply:
Ah makes sense, thanks for the change